tools/functional-tester/etcd-tester/lease_stresser.go

```diff
@@ -0,0 +1,300 @@
+	var wg sync.WaitGroup
+	wg.Add(neededLeases)
+	for i := 0; i < neededLeases; i++ {
+		go func() {
+			defer wg.Done()
+			leaseID, err := ls.createLeaseAndKeepAlive(ctx)
+			if err != nil {
+				plog.Errorf("lease creation error: (%v)", err)
+				return
+			}
+			plog.Debugf("lease %v created ", leaseID)
+			ls.attachKeysWithLease(ctx, leaseID)
+			ls.mu.Lock()
+			ls.aliveLeases[leaseID] = struct{}{}
+			ls.mu.Unlock()
+		}()
+	}
+	wg.Wait()
+}
+
+func (ls *leaseStresser) randomlyDropLeases(ctx context.Context) {
+	var wg sync.WaitGroup
+	for l := range ls.aliveLeases {
+		wg.Add(1)
+		go func(leaseID int64) {
+			defer wg.Done()
+			if !ls.randomlyDropLease(ctx, leaseID) {
+				return
+			}
+			ls.mu.Lock()
+			plog.Debugf("lease %v dropped ", leaseID)
+			ls.revokedLeases[leaseID] = struct{}{}
+			delete(ls.aliveLeases, leaseID)
+			ls.mu.Unlock()
+		}(l)
+	}
+	wg.Wait()
+}
+
+func (ls *leaseStresser) getLeaseByID(ctx context.Context, leaseID int64) (*pb.LeaseTimeToLiveResponse, error) {
+	ltl := &pb.LeaseTimeToLiveRequest{
+		ID:   leaseID,
+		Keys: true,
+	}
+
+	return ls.lc.LeaseTimeToLive(ctx, ltl, grpc.FailFast(false))
+}
+
+func (ls *leaseStresser) hasLeaseExpired(ctx context.Context, leaseID int64) (bool, error) {
+	resp, err := ls.getLeaseByID(ctx, leaseID)
+	plog.Debugf("hasLeaseExpired %v resp %v error (%v)", leaseID, resp, err)
+	if rpctypes.Error(err) == rpctypes.ErrLeaseNotFound {
+		return true, nil
+	}
+	return false, err
+}
+
+// The keys attached to the lease has the format of "<leaseID>_<idx>" where idx is the ordering key creation
+// Since the format of keys contains about leaseID, finding keys base on "<leaseID>" prefix
+// determines whether the attached keys for a given leaseID has been deleted or not
+func (ls *leaseStresser) hasKeysAttachedToLeaseExpired(ctx context.Context, leaseID int64) (bool, error) {
+	resp, err := ls.kvc.Range(ctx, &pb.RangeRequest{
+		Key:      []byte(fmt.Sprintf("%d", leaseID)),
+		RangeEnd: []byte(clientv3.GetPrefixRangeEnd(fmt.Sprintf("%d", leaseID))),
+	}, grpc.FailFast(false))
+	plog.Debugf("hasKeysAttachedToLeaseExpired %v resp %v error (%v)", leaseID, resp, err)
+	if err != nil {
+		plog.Errorf("retriving keys attached to lease %v error: (%v)", leaseID, err)
+		return false, err
+	}
+	return len(resp.Kvs) == 0, nil
+}
+
+func (ls *leaseStresser) createLeaseAndKeepAlive(ctx context.Context) (int64, error) {
+	leaseID, err := ls.createLease(ctx)
+	if err != nil {
+		return -1, err
+	}
+	// keep track of all the keep lease alive go routines
+	ls.aliveWg.Add(1)
+	go ls.keepLeaseAlive(ctx, leaseID)
+	return leaseID, nil
+}
+
+func (ls *leaseStresser) createLease(ctx context.Context) (int64, error) {
+	resp, err := ls.lc.LeaseGrant(ctx, &pb.LeaseGrantRequest{
+		TTL: TTL,
+	})
+	if err != nil {
+		return -1, err
+	}
+	return resp.ID, nil
+}
+
+func (ls *leaseStresser) keepLeaseAlive(ctx context.Context, leaseID int64) {
+	defer ls.aliveWg.Done()
```

Comment:
i removed it from the cancell(), right now i am unsure where to put ls.aliveWg.Wait() since I don't want to cancel the stresser after each round. The `ls.aliveWg.Wait()` should be only called when `cluster.terminate()` is invoked.